### PR TITLE
Fix issue in get_temp_export_dir under python 3

### LINKED ISSUE #21403: Crash during folder creation from Estimator exporter.py in Python 3.6
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.8.0 and 1.9.0
- **Python version**: 3.6
- **GPU model and memory**: GTX 1080 Ti
- **Exact command to reproduce**: See [here](https://github.com/tensorflow/models/issues/4996#issue-347431961)

### Describe the problem
When exporting of a snapshot during training with Estimator, [`export.py`](https://github.com/tensorflow/tensorflow/blob/e70f94ee089abbb9eb70361b5cdef55aa9beb18b/tensorflow/python/estimator/export/export.py#L600) creates an invalid path with Python 3.6, due to handling of strings as bytes (!!!). 

In [`estimator.py`](https://github.com/tensorflow/tensorflow/blob/e70f94ee089abbb9eb70361b5cdef55aa9beb18b/tensorflow/python/estimator/estimator.py#L779), the following string is generated for the variable `export_dir`: `b'C:/Users/Alex/pet-train\\export\\Servo\\1533549311'`, which is afterwards corrupted by the `get_temp_export_dir` command, which defaces it into `b"C:/Users/Alex/pet-train\\export\\Servo\\temp-b'1533549311'"` which further down the road causes the program to crash with the following exception:

```
Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm 2018.1.2\helpers\pydev\pydevd.py", line 1664, in <module>
    main()
  File "C:\Program Files\JetBrains\PyCharm 2018.1.2\helpers\pydev\pydevd.py", line 1658, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "C:\Program Files\JetBrains\PyCharm 2018.1.2\helpers\pydev\pydevd.py", line 1068, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm 2018.1.2\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:/Users/Alex/Repositories/MusicObjectDetector-TF2/research/object_detection/model_main.py", line 101, in <module>
    tf.app.run()
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\platform\app.py", line 126, in run
    _sys.exit(main(argv))
  File "C:/Users/Alex/Repositories/MusicObjectDetector-TF2/research/object_detection/model_main.py", line 97, in main
    tf.estimator.train_and_evaluate(estimator, train_spec, eval_specs[0])
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\estimator\training.py", line 439, in train_and_evaluate
    executor.run()
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\estimator\training.py", line 518, in run
    self.run_local()
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\estimator\training.py", line 657, in run_local
    eval_result = evaluator.evaluate_and_export()
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\estimator\training.py", line 858, in evaluate_and_export
    self._export_eval_result(eval_result, is_the_final_export)
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\estimator\training.py", line 889, in _export_eval_result
    is_the_final_export=is_the_final_export)
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\estimator\exporter.py", line 177, in export
    is_the_final_export)
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\estimator\exporter.py", line 123, in export
    strip_default_attrs=self._strip_default_attrs)
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\estimator\estimator.py", line 644, in export_savedmodel
    builder = saved_model_builder.SavedModelBuilder(temp_export_dir)
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\saved_model\builder_impl.py", line 92, in __init__
    file_io.recursive_create_dir(self._export_dir)
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\lib\io\file_io.py", line 374, in recursive_create_dir
    pywrap_tensorflow.RecursivelyCreateDir(compat.as_bytes(dirname), status)
  File "C:\Programmieren\Anaconda3\lib\site-packages\tensorflow\python\framework\errors_impl.py", line 519, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.NotFoundError: Failed to create a directory: C:/Users/Alex/pet-train\export\Servo\temp-b'1533549311'; No such file or directory

Process finished with exit code -1
```

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #21403 where
tensorflow.python.estimator.export.get_temp_export_dir
does not work correctly under python 3 due to the combination
of bytes vs non-bytes concatenation:
```
Python 3.5.2 (default, Nov 23 2017, 16:37:01)
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from tensorflow.python.estimator.export import export as export_helpers
>>> export_dir = export_helpers.get_timestamped_export_dir(b"/a/b/")
>>> export_dir
b'/a/b/1533609814'
>>> export_helpers.get_temp_export_dir(export_dir)
b"/a/b/temp-b'1533609814'"
>>>
```
This fix address the issue by adding compat.as_bytes to 'temp-'
before concatenation.

This fix fixes #21403.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by rohan100jain** [comment]
Can you add some test for the fix?

**Comment by yongtang** [comment]
Thanks @rohan100jain for the review. I updated the PR with unit test added. Without the change introduced in this PR the test will fail under python 3 (works fine in python 2). Please take a look and let me know if there are any issues.

File: tensorflow/python/estimator/export/export.py [hunk 1]
```diff
@@ -589,5 +589,6 @@ def get_temp_export_dir(timestamped_export_dir):
   """
   (dirname, basename) = os.path.split(timestamped_export_dir)
   temp_export_dir = os.path.join(
-      compat.as_bytes(dirname), compat.as_bytes('temp-{}'.format(basename)))
+      compat.as_bytes(dirname),
+      compat.as_bytes('temp-') + compat.as_bytes(basename))
```
Comment (superbobry) [inline tensorflow/python/estimator/export/export.py]: Consider using byte literals instead of manual conversion: b"temp-"

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
Sorry this fell through the cracks. @yongtang could you pull rebase and push again?

**Pull request closed without merging**